README.md
# Browser-Telemetry

`browser-telemetry` module collects client(browser) side errors,logs,metrics(timing & navigation), uncaught exceptions etc and sends back to server for logging & alerting.

## Features
* Intercept console.log/error/info/warn on browser
* Intercept Uncaught Exceptions on browser
* Batch update
* Timing & Navigation metrics
* Sampling (Browser side & Server Side)
* Custom Data Provider plugins

## TBD
* Rate Limiting
* Token based API access
* CORS

## Usage

### Environment
* Minimal **Node Version:** **v8.0.0**

### Client Side Usage

```html
    <script src="static/logger.min.js">
    </script>
```
Minified JS **`logger.min.js`**

**Example:**
```html
<html>
    <head>
        <script src="static/logger.min.js">
        </script>

        <script>
            function logData() {
                $logger.init({ 'url': '/api/log', 'flushInterval': 1000, 'samplingRate': 50, 'sendMetrics': true});

                $logger.registerPlugin('custom', function(payload) {
                    payload.custom = {
                        'Pagename': 'HomePage'                        
                    };
                });
                $logger.log('Hello, Logging Data!!');

                //Console
                console.log('Hello, from console.log');
                console.error('Hello, from console.error');
                console.info('Hello, from console.info');

                //Client Side Uncaught Exception
                throw new Error('Uncaught Error');
            }
        </script>
    </head>
    <body onload="logData()">
            <h1>Hello World!!</h1>
    </body>

</html>
```

### Server Side Usage

```javascript
    
    let app = express();
    const path = path.resolve('browser-telemetry');
    app.use('/static', express.static(path.join(path, '../browser')));
    app.use(bodyParser.json({ type: 'application/*+json' }));
    app.use(bodyParser.json({ type: 'text/plain' }));

    //require Logger
    let loggerMiddleware = require('./');
    
    //Add Logger Middleware
    app.use(loggerMiddleware({
        path: '/api/log',
        log: function(req, payload) {
            console.log('Metrics:', payload.metrics);   
            
            //Consoles from Client Side            
            payload.logs.forEach((event) => {
                console.log(`${event.level}: ${JSON.stringify(event.msg)}`);
            });   
            console.log('Custom Plugin Data:', payload.custom);   
        }
    }));
```
### Payload Data
* **logs/errors:** ALL console logs/errors are intercepted and sent in `logs` field.
* **Uncaught Exceptions:** ALL uncaught exceptions are intercepted and sent in `logs` field as `ERROR`.
* **Metrics:** Browser load times are captured using timing & navigation API(Only in compatible browsers). 
    * **rc:** Redirect Count, number of times page is redirected before hitting current page.
    * **lt:** Load Time, load time of the page.
    * **ct:** Connect Time, time took to connect to server.
    * **rt:** Render Time, time took to render the page.
    * **navType:** Tells you how the page is accessed. `0=Navigate, 1=Reload, 2=History`

```json
{
  "metrics": {
    "navType": 0,
    "rc": 0,
    "lt": 255,
    "ct": 72,
    "rt": 147
  },
  "logs": [
    {
      "level": "LOG",
      "msg": "Hello, Logging Data!!"
    },
    {
      "level": "LOG",
      "msg": "Hello from console.log"
    },
    {
      "level": "ERROR",
      "msg": "Hello from console.error"
    },
    {
      "level": "INFO",
      "msg": "Hello from console.info"
    },
    {
      "level": "ERROR",
      "msg": "Uncaught Error: Uncaught http://localhost:8080/ 20 23 Error: Uncaught\n    at logData (http://localhost:8080/:20:23)\n    at onload (http://localhost:8080/:24:30)"
    }
  ],
  "custom": {
    "pageName": "HomePage"
  }
}
```

## API
### Client Side API

* #### $logger
    On load of Javascript file, `$logger` object gets hooked up to `window` object for global access.

* #### $logger.init(object)
    For initializing logger, call this API. 
    **Input:**
    ```javascript
    {
        "url": "api/log", //Relative path
        "flushInterval": 1000, //1sec,
        "samplingRate": 10, //10%, Client Side Sampling
        "isInSampling": true, //Flag from Server Side Sampling
        "sendMetrics": true, //Flag to send metrics or not
        "logLevels": ["info", "log", "debug", "error", "warn"] //List of enabled console methods
    }
    ```

* #### $logger.registerPlugin(pluginName, callback)
    * **callback(payload)**
    Some times you need to send your own custom data. You can attach custom data or transform data by registering your own plugin. 
    Payload object will be passed, which can be mutated in callback.

    On every flush, ALL registered plugins gets called for data transformation.

    ```javascript
        $logger.registerPlugin('custom', function(payload) {
            payload.Pagename = 'HomePage';
        });
    ```

### Server Side API

* #### Middleware
    #### require('browser-telemetry')(options)
    * **options:** 
        * **path**: Path on which logger should be mounted. This is the end where events from browser are posted.
        * **log**: A callback function which will be invoked on every event from client side.
            * **request**: Holds HTTP request object
            * **payload**: A payload which holds events from browser.

        ```javascript
            const loggerMiddleware = require('browser-telemetry');
            let app = reqiure('express');
            ...
            app.use(loggerMiddleware({
                path: 'path/to/mount',
                log: function(req, payload) {
                    console.log(payload);
                }
            }));
        ```

* #### Log Hook
    As an app developer, you can intercept browser events by hooking to `log` callbacks. Simply pass your custom function while registering middleware as shown below.
    
    Browser events are populated in **payload** param.

    ```javascript
        app.use(loggerMiddleware({
            path: 'path/to/mount',
            log: function(req, payload) {
                console.log(payload);
            }
        }));
    ```
### File Size

Main motivation for creating this module is to reduce file size. Currently, minified client side JS file size is **~2KB**.

### Example
See the working example in `example` folder.

**Running Example Server**
```shell
    node example/server.js

    //Open Browser pointing to http://localhost:8080
    //ALl Logs/Metrics will be dumped on Server side console.
```

### Ackowledgement

This module is created as an inspiration from [beaver-logger](https://github.com/krakenjs/beaver-logger). Main motivation is to reduce client side JS file size and provide minimal functionality for intercepting logs/metrics/uncaught exceptions on browser side.

### Privacy Notice

On running this module/Javascript file on browser, it collects page load metrics, console logs, error logs, unhandled exceptions etc and sends data to backend server for processing.

### License

Copyright 2018 eBay Inc.
 
Use of this source code is governed by an MIT-style license that can be found in the LICENSE file or at https://opensource.org/licenses/MIT.

CHANGELOG.md
# ChangeLog

## v1.0.0
* Initial Version

test/index-test.js
/**
 * Copyright (c) 2018 eBay Inc.
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 *  
 **/

'use strict';
const loggerMiddleware = require('../');
const request = require('request');
const async = require('async');

const assert = require('assert');
const Browser = require('zombie');

describe(__filename, function() {
    const app = require('./fixtures/app');
    var srv;
    before((cb)=> {
        srv = app.listen(9099, () => {
            console.log('Server started...');
            cb();
        });
    });
    
    after((cb)=> {
        srv && srv.close();
        setTimeout(cb, 2000);
    });

    it('should post metrics to server hook and invoke callback hook', function(done) {
        let payload = {"metrics":{"navType":"1","rc":"0","lt":"1115","ct":"370","rt":"1042"},
        "logs":[{"level":"LOG","msg":"Hello, $logger.log"},{"msg":"Hello,console.log"},{"level":"INFO","msg":"Hello,console.info"},{"level":"WARN","msg":"Hello,console.warn"},{"level":"DEBUG","msg":"Hello,console.debug"},{"level":"ERROR","msg":"Some Error Error: Error Object\n    at logData (http://localhost:8080/:17:45)\n    at onload (http://localhost:8080/:19:597)"},{"level":"ERROR","msg":"Uncaught SyntaxError: Unexpected token < in JSON at position 0 http://localhost:8080/ 1 1 SyntaxError: Unexpected token < in JSON at position 0\n    at JSON.parse (<anonymous>)\n    at XMLHttpRequest.t.onload"}],
        "custom":{"pageName":"HomePage"}};
        
        process.on('bowserPayload', (bowserPayload)=> {
            payload.url = 'http://localhost:9099/';
            payload.userAgent = 'Chrome';
            assert.equal(JSON.stringify(payload), bowserPayload, 'Payload should be same');
            process.removeAllListeners(['bowserPayload']);
            done();
        });

        request.post({url: 'http://localhost:9099/api/log', headers: {'user-agent': 'Chrome', 'referer': 'http://localhost:9099/'},form: payload}, function(err, res, body) {
            // payload.url = 'http://localhost:9099/';
            // payload.userAgent = 'Chrome';
            // assert.equal(JSON.stringify(payload), res.body, 'Payload should be same');
            assert.equal(200, res.statusCode, 'Should send 200 status code');
            // done();
        });                  
    });

    it('should load JS on Browser and fire beacon', function(done) {
                
        process.on('bowserPayload', (bowserPayload)=> {
            var pload = {"logs":[{"level":"LOG","msg":"Hello, Logging Data!!"},{"level":"LOG","msg":"Hello from console.log"},{"level":"ERROR","msg":"Hello from console.error"},{"level":"INFO","msg":"Hello from console.info"}],"custom":{"pageName":"HomePage"},"userAgent":"MSIE 9"};
            assert.equal(bowserPayload, JSON.stringify(pload));
            process.removeAllListeners(['bowserPayload']);
            done();
        });

        Browser.localhost('www.test.ebay.com', 9099);
            var browser = new Browser();
            browser.userAgent = 'MSIE 9';
            browser.debug = true;
            browser.on('xhr', function (event) {
                console.log('* XHR event: ', event);
            });
            browser.on('console', function(level, message) {
                console.log(level, message);                
            });
            browser.on('error',function (err){
               console.log('* error: ', err);
            });            
            browser.visit('/', function startXhr() {
                console.log('Main page is loaded');                                  
            });                    
    });

    it('should load JS on Browser and Log Only ERROR', function(done) {
                
        process.on('bowserPayload', (bowserPayload)=> {
            var pload = {"logs":[{"level":"ERROR","msg":"Hello from console.error"}],"custom":{"pageName":"HomePage"},"userAgent":"MSIE 9"};
            assert.equal(bowserPayload, JSON.stringify(pload));
            process.removeAllListeners(['bowserPayload']);
            done();
        });

        Browser.localhost('www.test.ebay.com', 9099);
            var browser = new Browser();
            browser.userAgent = 'MSIE 9';
            browser.debug = true;
            browser.on('xhr', function (event) {
                console.log('* XHR event: ', event);
            });
            browser.on('console', function(level, message) {
                console.log(level, message);                
            });
            browser.on('error',function (err){
               console.log('* error: ', err);
            });            
            browser.visit('/loglevels', function startXhr() {
                console.log('Main page is loaded');                                  
            });                    
    });

    it('should NOT fire Telemetry JS on Browser', function(done) {
                
        process.on('bowserPayload', (bowserPayload)=> {
            assert.fail('Should NOT Sent Events to Server');            
        });

        Browser.localhost('www.test.ebay.com', 9099);
            var browser = new Browser();
            browser.userAgent = 'MSIE 9';
            browser.debug = true;
            browser.on('xhr', function (event) {
                assert.fail('Should NOT call XHR Events');  
                // console.log('* XHR event: ', event);
            });
            browser.on('console', function(level, message) {
                console.log(level, message);                
                if(message.indexOf('Is in Sample') > -1) {
                    assert.equal(message, 'Is in Sample:  false');
                    setTimeout(()=> {
                        process.removeAllListeners(['bowserPayload']);
                        done();
                    }, 1000);                    
                }                 
            });
            browser.on('error',function (err){
               console.log('* error: ', err);
            });            
            browser.visit('/samplingoff', function startXhr() {
                console.log('Main page is loaded');                                  
            });                    
    });
});

test/fixtures/app.js
const express = require('express');
const bodyParser = require('body-parser');
const path = require('path');
const loggerMiddleware = require('../..');
const app = express();

app.use('/static', express.static(path.join(__dirname,'../..','browser')))
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json({ type: 'application/*+json' }));
app.use(bodyParser.json({ type: 'text/plain' }));
app.use(bodyParser.json({ type: 'text/html' }));

app.use(loggerMiddleware({
    path: '/api/log',
    log: function(req, payload) {
        process.emit('bowserPayload', JSON.stringify(payload));       
    }
}));

app.get('/samplingoff', (req, res)=> {    
    res.sendFile(path.join(__dirname, 'templates/sampling.html'), {
        headers: {
            'Content-Type': 'text/html'
        }
    });        
});

app.get('/loglevels', (req, res)=> {    
    res.sendFile(path.join(__dirname, 'templates/loglevels.html'), {
        headers: {
            'Content-Type': 'text/html'
        }
    });        
});

app.get('/', (req, res)=> {    
    res.sendFile(path.join(__dirname, 'templates/allfeatures.html'), {
        headers: {
            'Content-Type': 'text/html'
        }
    });    
});

module.exports = app;

test/fixtures/server.js
let express = require('express');
let path = require('path');
let bodyParser = require('body-parser');

let logger = require('./');

let loggerOptions = {
    path:  '/api/log',
    log: function(req, payload) {
        console.log('Inside Override', payload.metrics, payload.url, payload.ebay.rlogid, payload.ebay.pageName);
        let eventList = payload.logs;                

        eventList.forEach((event) => {
            console.log(`Override:: ${event.level}: ${JSON.stringify(event.msg)}`);
        }); 
    }
};

let app = express();

app.use(bodyParser.json({ type: 'application/*+json' }));
app.use(bodyParser.json({ type: 'text/plain' }));

// app.use('/browserstatic', express.static(path.join(require.resolve('browser-telemetry'), '..','browser')));
app.use('/static', express.static('browser'))
console.log('******',path.join(require.resolve('browser-telemetry'), '..','browser'));
app.use(logger(loggerOptions));

app.get('/', (req, res)=> {
    res.sendFile(path.join(__dirname, 'home.html'), {
        headers: {
            'Content-Type': 'text/html'
        }
    });
});

app.listen(8080);

lib/index.js
/**
 * Copyright (c) 2018 eBay Inc.
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 *  
 **/

let express = require('express');
let _ = require('underscore');
const tryRequire = require('try-require');

/**
 * Default Config on server side.
**/
const DEFAULT_CONFIG = {
    path: '/api/log',
    log: (req, payload) => {
        if(payload.metrics) {
            console.log('Metrics:', payload.metrics); 
        }        
        payload.logs.forEach((event) => {
            console.log(`${event.level}: ${JSON.stringify(event.msg)}`);
        }); 
    }
};

/**
 * Express Middleware for hooking URL and receiving client payload.
 * Calls Custom hook registered in middleware.
**/
function loggerMiddleware(options={}) {     
    _.defaults(options, DEFAULT_CONFIG);

    if(typeof options.log === 'string') {
        let logFile = options.log;
        options.log = tryRequire(options.log);
        if(!options.log) {
            console.error(`Unable to load the file ${logFile}, setting default callback.`);
            options.log = DEFAULT_CONFIG.log;
        }
    }

    let loggerApp = express();   
        
    loggerApp.post(options.path, (req, res, next)=> {
        try {
            let payload = req.body;
            payload.url = req.headers['referer'];
            payload.userAgent = req.headers['user-agent'];

            options.log(req, payload);
            res.status(200).end();

        } catch(error) {
            console.error(error);
            res.status(500).end();
        }
        
    });
    return loggerApp;
}

module.exports = loggerMiddleware;

browser/logger.js
/**
 * Copyright (c) 2018 eBay Inc.
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 *  
 **/

/**
 * Default configuration
**/
var _DEFAULTS = {   
                    'url': '/api/log', 
                    'flushInterval': 1000, 
                    isInSampling: true, 
                    samplingRate: 100,
                    collectMetrics: true,
                    logLevels: ['log', 'info', 'warn','debug','error']
                };

/**
 * Logger Class which exposes API for intercepting log, errors & collects metrics
**/
function Logger() {   
    this.buffer = []; 
    this.plugins = {};

    this.url = _DEFAULTS.url;
    this.flushInterval = _DEFAULTS.flushInterval;
    this.collectMetrics = _DEFAULTS.collectMetrics;
    this.logLevels = _DEFAULTS.logLevels;    
}

/**
 * Init API for intializing the class with paramaters.
**/
Logger.prototype.init = function(options) {
    var thisObj = this;
    options = options || _DEFAULTS;    
    this.url = options.url || this.url;
    this.flushInterval = options.flushInterval || this.flushInterval; //In ms
    this.logLevels = options.logLevels || this.logLevels;

    this.collectMetrics = options.collectMetrics !== undefined ? options.collectMetrics : this.collectMetrics;

    //Use Sampling Flag provide in init() or calculate Sampling factor based on Sampling Rate
    var isInSampling = options.isInSampling !== undefined ? options.isInSampling : sample(options.samplingRate);
    console.log('Is in Sample: ', isInSampling);
    var _this = this;

    //Setup timer & flush ONLY if this is in Sampling
    if(isInSampling) {
        var loglevels = ['log', 'info', 'warn','debug','error'];        

        loglevels.forEach(function(level) {            
                var _fn = console[level];
                console[level] = function() {
                    var args = Array.prototype.slice.call(arguments);
                    _this[level](args);
                    _fn.apply(console, args);
                }
        });

        setInterval(function() {
            if(thisObj.buffer.length > 0) {
                thisObj.flush();
            }                
        }, options.flushInterval);
    }
}

/**
 * API for registering custom functions.
**/
Logger.prototype.registerPlugin = function(property, customFunction) {
    this.plugins[property] = customFunction;
}

/**
 * Collects metrics using navigation API 
**/
Logger.prototype.metrics = function() {
    if(!(window && window.performance)) {
        return;
    }

    var perf = window.performance;
    var perfData = perf.timing;
    var navData = perf.navigation;
    var metrics = {
        'navType': navData.type, // 0=Navigate, 1=Reload, 2=History
        'rc': navData.redirectCount,
        'lt': perfData.loadEventEnd - perfData.navigationStart, //PageLoadTime
        'ct': perfData.responseEnd - perfData.requestStart, //connectTime
        'rt': perfData.domComplete - perfData.domLoading //renderTime
    };
    return metrics;
}

Logger.prototype.log = function() {
    this.addToQ('LOG', arguments);
}

Logger.prototype.info = function() {
    this.addToQ('INFO', arguments);
}

Logger.prototype.debug = function() {
    this.addToQ('DEBUG', arguments);
}

Logger.prototype.warn = function() {
    this.addToQ('WARN', arguments);
}

Logger.prototype.error = function() {
    var args = [];
    Array.prototype.slice.call(arguments[0]).forEach(function(elem) {
        
        if(typeof elem === 'object' && elem.stack) {
            args.push(elem.stack);
        } else {
            args.push(elem);
        }
    });

    this.addToQ('ERROR', args);
}

/**
 * Clears Buffer
**/
Logger.prototype.clearBuffer = function(clearFromIndex) {
    this.buffer = this.buffer.slice(clearFromIndex);
}

/**
 * Adds message and type to Queue
**/
Logger.prototype.addToQ = function(type, args) {
    if(this.logLevels.indexOf(type) > -1 || this.logLevels.indexOf(type.toLowerCase()) > -1) {
        var message = (args.length>0 && [].join.call(args, ' ')) || '';
        this.buffer.push({level: type, msg: message});
    }
}

/**
 * Flushes data from buffer
**/
Logger.prototype.flush = function() {
    var _this = this;

    if(_this.buffer.length < 1) {
        return;
    } 
    var bufSize = _this.buffer.length;
    var payload = {
        'metrics': _this.metrics(),
        'logs': _this.buffer        
        
    };

    Object.keys(_this.plugins).forEach(function(property) {
        _this.plugins[property](payload);
    });

    if(navigator && navigator.sendBeacon) {
        var status = navigator.sendBeacon(_this.url, JSON.stringify(payload));
        if(status) {
            _this.clearBuffer(bufSize);
        }        
    } else {        
        var xhr = new XMLHttpRequest();
        xhr.open('POST', this.url, true); // third parameter indicates sync xhr
        xhr.setRequestHeader('Content-Type', 'text/plain');
        xhr.onreadystatechange = function() {//Call a function when the state changes.
            if(xhr.readyState == 4 && xhr.status == 200) {
                // Request finished. Do processing here.
                _this.clearBuffer(bufSize);
            }            
        }
        xhr.send(JSON.stringify(payload));
    }
}

/**
 * Client side Sampling API
**/
function sample(samplingRate) {
    if(Math.random() * 100 < samplingRate) {
        return true;
    } else {
        return false;
    }    
}

function intialize() {    
    var logger = new Logger();
    if(window) {
        window.$logger = logger;
        
        var _onerror = window.onerror;
        //Handle Uncaught Errors
        window.onerror = function() {
            var args = Array.prototype.slice.call(arguments);
            logger.error(args);
            if(_onerror) {
                return _onerror.apply(window, args);
            } else {
                return false;
            }            
        };        
    }    
}

intialize();